### DISCUSSION TITLE
6 AXIS CNC

### DISCUSSION THREAD

**Discussion by alphaonex86** [fra589/grbl-Mega-5X#359] [Q&A]
Hi,
I have my own driver, 1X, 5Y, 3Z, all with 2+ optical end-stop for each AXIS (22 end stop with duplicate AXIS). Then the other AXIS is linear, not rotatory. Each Y it's for 1 tools, 3Y it's with spindle motor.
All my driver have error output is case of error detection.
What I need change into the source to have full working/homing and take care of error?
Sorry for my bad English.

**Comment by fra589** [comment]
Hi @alphaonex86,  
  
If I have a good count: 1x + 5y + 3z = 9 logicals axis, with duplicat axes, you have 2 end stops per axis, so: 22 / 2 = 11 reals axes !!!

grbl-Mega-5X is only able to drive 6 reals axes. The Arduino mega 2560 board doesn't have the hardware capacity to do more than this.

To drive the huge number of axes of your project, you will need to use 2 or more cards with grbl-Mega-5X. So, you will need to add synchronization communication between them. Perhaps this can be done directly in GCode using the digital inputs/outputs which grbl-Mega-5X is able to drive ([M62-M65] Digital output, the digital input synchronization GCode (M66) in not yet implemented in gcode.c).

Moreover, there is no input to drive error input message from drivers.
Depending of the error output format of your drivers, this may be more or less complicated...
If the output is only digital signal up or down, this can be driven by the digital input functionality. If it's a serial communication, you will need to write a new module for a second serial port com.

@++;
Gauthier.

**Comment by alphaonex86** [comment]
I search detect the bed leveling, but not as square, following the path and probe (some zone is isolated, other is conductive). I can generate the GCode, I need just have GCode way to detect the level (I can connect opto + electric contact detector).

Any GCode to just detect and get the level? pin to connect the sensor?

**Comment by fra589** [comment]
Hi @alphaonex86,  
  
The GCode `G38.n` (_with n=2, 3, 4 or 5_) probe functionality is that you need.
See http://linuxcnc.org/docs/html/gcode/g-code.html#gcode:g38 for this functionality reference.

When you issue a G38.n, Grbl reply with a PRB: probe parameter message which give the axes values of contact point and includes an additional : suffix value (boolean) which denotes whether the last probe cycle was successful or not.

The last probe value can also be retrieved with the `$#` command (see https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-v1.2-interface for output format)

You can see on wiki (https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout) than the default probe pin is on A15.

@++;
Gauthier.

**Reply by alphaonex86** [reply]
thanks for your help

I have resolved the sync AXIS problem, I'm developing the software for my crazy CNC, (optical calibration + electric), API to accept order from web page, error, security, ... from gerber to gcode generation + validation. Lot of cool stuff.

**Comment by alphaonex86** [comment]
For now I failed on:
- End stop, not stop when the end stop is activated
- Homing, I need do manual homing only when I trigger it
I have correctly wire 2 end Stop and is correctly detected by digitalRead of arduino MEGA

**Reply by fra589** [reply]
!?!?!?
More precises info are needed to understand your problem :-(
- config.h, cpu_map.h & default.h: not modified or some adjustment?
- $$ output,
- $I output,
- "not stop when the end stop is activated": when homing? with normal move?
=> See cpu_map.h, line 200 to 206... Is `#define ENABLE_RAMPS_HW_LIMITS` enabled ?
- "Homing, I need do manual homing only when I trigger it":
  - what do you mean by "doing manual homing"?
  - what do you mean by "only when I trigger it"?
  - what are your HOMING_CYCLEs definitions in config.h?
  - what error, or what behavior when commanding $H, $H<axis>?
  
In computer programming, it is impossible for me to debug the specific error: "It doesn't work"

@++;
Gauthier.

**Comment by alphaonex86** [comment]
- config.h, cpu_map.h & default.h: not modified

- ```
>>> $I
[VER:1.2h.20220109:]
[AXS:5:XYZAB]
[OPT:VNMGH,35,255,120]
ok
>>> $$
$0 = 10    (Step pulse time, microseconds)
$1 = 254    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 0    (Invert probe pin, boolean)
$10 = 1    (Status report options, mask)
$11 = 0.020    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 1    (Soft limits enable, boolean)
$21 = 1    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 0    (Homing direction invert, mask)
$24 = 25.000    (Homing locate feed rate, mm/min)
$25 = 250.000    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 5.000    (Homing switch pull-off distance, millimeters)
$30 = 12000    (Maximum spindle speed, RPM)
$31 = 550    (Minimum spindle speed, RPM)
$32 = 0    (Laser-mode enable, boolean)
$100 = 20.000    (X-axis travel resolution, step/mm)
$101 = 400.000    (Y-axis travel resolution, step/mm)
$102 = 400.000    (Z-axis travel resolution, step/mm)
$103 = 8.889   
$104 = 8.889   
$110 = 750.000    (X-axis maximum rate, mm/min)
$111 = 750.000    (Y-axis maximum rate, mm/min)
$112 = 750.000    (Z-axis maximum rate, mm/min)
$113 = 1440.000   
$114 = 1440.000   
$120 = 50.000    (X-axis acceleration, mm/sec^2)
$121 = 50.000    (Y-axis acceleration, mm/sec^2)
$122 = 50.000    (Z-axis acceleration, mm/sec^2)
$123 = 50.000   
$124 = 50.000   
$130 = 400.000    (X-axis maximum travel, millimeters)
$131 = 200.000    (Y-axis maximum travel, millimeters)
$132 = 200.000    (Z-axis maximum travel, millimeters)
$133 = 360.000   
$134 = 180.000   
ok
- not stop when the end stop is activated, with normal move
- what are your HOMING_CYCLEs definitions in config.h? -> as default, I have only X/Y or X/Y/Z configuration.
- I have try to play with $20 $21 $22, I think I need $20=0 $21=1 $22=0 but is this case I start with alarm, 0V/LOW whenis NOT at top
`<Alarm|MPos:0.000,0.000,0.000,0.000,0.000|FS:0,0|Pn:YZAB|WCO:-100.200,-200.000,0.000,0.000,0.000>
<Alarm|MPos:0.000,0.000,0.000,0.000,0.000|FS:0,0|Pn:YZAB|Ov:100,100,100>` with $21=1```

**Reply by alphaonex86** [reply]
INVERT_MIN_LIMIT_PIN_MASK/INVERT_MAX_LIMIT_PIN_MASK Found, now I just need it stop when reach the min or max,

**Comment by fra589** [comment]
Hello @alphaonex86,  
  
Grbl (as many other CNC firmware using stepper motors) is not able to know it's phisical position at power on. So, if you enable limits (software or hardware ($20 / $21)) you also need to enable homing ($22). Then, at power on, Grbl come in Alarm status and wait for a homing to know it's position and clear this alarm status. You can clear this Alarm status by issuing the `$X` command, but if you do this, Grbl will not be able to know it's real position and it will can't drive it's software limits correctly.

By default, hardware limits are not enabled in grbl-Mega-5X since this lower it's performance a little bit. To enable hardware limits, see cpu_map.h: line 200 to 206... and uncomment the line (remove the // at begining)
`#define ENABLE_RAMPS_HW_LIMITS`
With this (and config $21=1), when a limit switch will be trigered during move, Grbl will suddenly stop (loosing it's phisical position because of no deceleration driving) and come to alarm 1. You will need to reset and re-homing the machine to clear this alarm state (or issue `$X` with the same problem than at power on).
I rather advise instead to correctly configure the dimensions of the machine (max travels) so that the software limits will work well. With software limits, Grbl don't loose it's phisical position in case of limit alarm.
  
@++;  
Gauthier.

**Reply by alphaonex86** [reply]
Thanks, as you know, first try on CNC will fail, it's more than all for this case, when I'm sure all is working, I will disable the hardware limit.

Here stepper motor cost over 1500€/USD, then I have to protect it at all cost during the development. The importation price in Bolivia is at 1000USD+300% of the goods price.
